Sources/SwiftGRPCNIO/GRPCTLSVerificationHandler.swift

```diff
@@ -0,0 +1,71 @@
+import Foundation
+import NIO
+import NIOSSL
+import NIOTLS
+
+/// Application protocol identifiers for ALPN.
+public enum GRPCApplicationProtocolIdentifier: String, CaseIterable {
+  // This is not in the IANA ALPN protocol ID registry, but may be used by servers to indicate that
+  // they serve only gRPC traffic. It is part of the gRPC core implementation.
+  case gRPC = "grpc-ext"
+  case h2 = "h2"
+}
+
+/// A helper `ChannelInboundHandler` to verify that a TLS handshake was completed successfully
+/// and that the negotiated application protocol is valid.
+///
+/// The handler holds a promise which is succeeded on successful verification of the negotiated
+/// application protocol and failed if any error is received by this handler or an invalid
+/// application protocol was negotiated.
+///
+/// Users of this handler should rely on the `verification` future held by this instance.
+///
+/// On fulfillment of the promise this handler is removed from the channel pipeline.
+public class GRPCTLSVerificationHandler: ChannelInboundHandler, RemovableChannelHandler {
```

Comment:
I assume this class is required _in addition to_ the `applicationProtocols: GRPCApplicationProtocolIdentifier.allCases.map { $0.rawValue }` above?

Reply:
Yes, that's correct. We also check that the handshake was successful here as well as the negotiated protocol.